ledger/complete/wal/checkpointer_test.go

```diff
@@ -366,9 +333,115 @@ func Test_Checkpointing(t *testing.T) {
 				require.Equal(t, values2[i], payloads5[i].Value)
 			}
 		})
+
+		t.Run("corrupted checkpoints are skipped", func(t *testing.T) {
+
+			f6, err := mtrie.NewForest(pathByteSize, dir, size*10, metricsCollector, func(tree *trie.MTrie) error { return nil })
+			require.NoError(t, err)
+
+			wal6, err := realWAL.NewWAL(zerolog.Nop(), nil, dir, size*10, pathByteSize, segmentSize)
+			require.NoError(t, err)
+
+			// make sure no earlier checkpoints exist
+			require.NoFileExists(t, path.Join(dir, "checkpoint.0000008"))
+			require.NoFileExists(t, path.Join(dir, "checkpoint.0000006"))
+			require.NoFileExists(t, path.Join(dir, "checkpoint.0000004"))
+
+			require.FileExists(t, path.Join(dir, "checkpoint.00000010"))
+
+			// create missing checkpoints
+			checkpointer, err := wal6.NewCheckpointer()
+			require.NoError(t, err)
+
+			err = checkpointer.Checkpoint(4, func() (io.WriteCloser, error) {
+				return checkpointer.CheckpointWriter(4)
+			})
+			require.NoError(t, err)
+			require.FileExists(t, path.Join(dir, "checkpoint.00000004"))
+
+			err = checkpointer.Checkpoint(6, func() (io.WriteCloser, error) {
+				return checkpointer.CheckpointWriter(6)
+			})
+			require.NoError(t, err)
+			require.FileExists(t, path.Join(dir, "checkpoint.00000006"))
+
+			err = checkpointer.Checkpoint(8, func() (io.WriteCloser, error) {
+				return checkpointer.CheckpointWriter(8)
+			})
+			require.NoError(t, err)
+			require.FileExists(t, path.Join(dir, "checkpoint.00000008"))
+
+			// corrupt checkpoints
+			randomlyModifyFile(t, path.Join(dir, "checkpoint.00000006"))
+			randomlyModifyFile(t, path.Join(dir, "checkpoint.00000008"))
+			randomlyModifyFile(t, path.Join(dir, "checkpoint.00000010"))
+
+			// make sure 10 is latest checkpoint
+			latestCheckpoint, err := checkpointer.LatestCheckpoint()
+			require.NoError(t, err)
+			require.Equal(t, 10, latestCheckpoint)
+
+			// at this stage, number 4 should be the latest valid checkpoint
+			// check other fail to load
+
+			_, err = checkpointer.LoadCheckpoint(10)
+			require.Error(t, err)
+			_, err = checkpointer.LoadCheckpoint(8)
+			require.Error(t, err)
+			_, err = checkpointer.LoadCheckpoint(6)
+			require.Error(t, err)
+			_, err = checkpointer.LoadCheckpoint(4)
+			require.NoError(t, err)
+
+			err = wal6.ReplayOnForest(f6)
```

Comment:
As long as this passes we consider this function to be :+1:?

Reply:
I would assume so, as we checked that checkpoints are indeed corrupted, but you are right, I can add extra check to make sure all data is there